File: frontend/src/stores/system.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { api } from "@/api";

export interface SystemInfo {
    hostname: string;
    os: string;
    os_version: string;
    kernel: string;
    uptime: number;
    cpu_count: number;
    total_memory: number;
}

export interface DiskInfo {
    name: string;
    mount_point: string;
    total: number;
    used: number;
    available: number;
    percent: number;
}

export interface NetworkInfo {
    name: string;
    rx_bytes: number;
    tx_bytes: number;
    rx_packets: number;
    tx_packets: number;
}

export interface SystemStats {
    cpu_usage: number;
    memory_used: number;
    memory_total: number;
    memory_percent: number;
    swap_used: number;
    swap_total: number;
    disks: DiskInfo[];
    networks: NetworkInfo[];
    load_avg: [number, number, number];
    timestamp: number;
}

export const useSystemStore = defineStore("system", () => {
    const info = ref<SystemInfo | null>(null);
    const stats = ref<SystemStats | null>(null);
    const statsHistory = ref<SystemStats[]>([]);
    const eventSource = ref<EventSource | null>(null);
    const isConnected = ref(false);
    const retryTimer = ref<ReturnType<typeof setTimeout> | null>(null);

    const fetchInfo = async () => {
        const response = await api.get("/system/info");
        info.value = response.data;
    };

    const fetchStats = async () => {
        const response = await api.get("/system/stats");
        stats.value = response.data;
        addToHistory(response.data);
    };

    const addToHistory = (stat: SystemStats) => {
        statsHistory.value.push(stat);
        // Keep last 60 data points (1 minute of data at 1s intervals)
        if (statsHistory.value.length > 60) {
            statsHistory.value.shift();
        }
    };

    const clearRetryTimer = () => {
        if (retryTimer.value) {
            clearTimeout(retryTimer.value);
            retryTimer.value = null;
        }
    };

    const startStreaming = () => {
        if (eventSource.value) return;
        clearRetryTimer();

        eventSource.value = new EventSource("/api/system/stats/stream");

        eventSource.value.onopen = () => {
            isConnected.value = true;
        };

        eventSource.value.onmessage = (event) => {
            const data = JSON.parse(event.data) as SystemStats;
            stats.value = data;
            addToHistory(data);
        };

        eventSource.value.onerror = () => {
            isConnected.value = false;
            stopStreaming();
            // Retry after 5 seconds
            retryTimer.value = setTimeout(() => {
                retryTimer.value = null;
                startStreaming();
            }, 5000);
        };
    };

    const stopStreaming = () => {
        clearRetryTimer();
        if (eventSource.value) {
            eventSource.value.close();
            eventSource.value = null;
            isConnected.value = false;
        }
    };

    return {
        info,
        stats,
        statsHistory,
        isConnected,
        fetchInfo,
        fetchStats,
        startStreaming,
        stopStreaming,
    };
});

File: backend/src/api/system.rs
use axum::{
    extract::State,
    response::sse::{Event, KeepAlive, Sse},
    routing::get,
    Json, Router,
};
use serde::Serialize;
use std::convert::Infallible;
use std::time::Duration;
use tokio_stream::{wrappers::IntervalStream, StreamExt};

use crate::AppState;

#[derive(Debug, Serialize)]
pub struct SystemInfo {
    pub hostname: String,
    pub os: String,
    pub os_version: String,
    pub kernel: String,
    pub uptime: u64,
    pub cpu_count: usize,
    pub total_memory: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct SystemStats {
    pub cpu_usage: f32,
    pub memory_used: u64,
    pub memory_total: u64,
    pub memory_percent: f32,
    pub swap_used: u64,
    pub swap_total: u64,
    pub disks: Vec<DiskInfo>,
    pub networks: Vec<NetworkInfo>,
    pub load_avg: [f64; 3],
    pub timestamp: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct DiskInfo {
    pub name: String,
    pub mount_point: String,
    pub total: u64,
    pub used: u64,
    pub available: u64,
    pub percent: f32,
}

#[derive(Debug, Clone, Serialize)]
pub struct NetworkInfo {
    pub name: String,
    pub rx_bytes: u64,
    pub tx_bytes: u64,
    pub rx_packets: u64,
    pub tx_packets: u64,
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/info", get(system_info))
        .route("/stats", get(system_stats))
        .route("/stats/stream", get(stats_stream))
}

async fn system_info(State(state): State<AppState>) -> Json<SystemInfo> {
    let info = state.monitor.get_system_info();
    Json(info)
}

async fn system_stats(State(state): State<AppState>) -> Json<SystemStats> {
    let stats = state.monitor.get_stats();
    Json(stats)
}

async fn stats_stream(
    State(state): State<AppState>,
) -> Sse<impl tokio_stream::Stream<Item = Result<Event, Infallible>>> {
    let interval = tokio::time::interval(Duration::from_secs(1));
    let stream = IntervalStream::new(interval).map(move |_| {
        let stats = state.monitor.get_stats();
        let json = serde_json::to_string(&stats).unwrap_or_default();
        Ok(Event::default().data(json))
    });

    Sse::new(stream).keep_alive(KeepAlive::default())
}
